### PR TITLE
Add support for HDInsight Cluster with Enterprise Security Package

### PR DESCRIPTION
**Pull request by neil-yechenwei**
Currently, TF HDInsight Cluster doesn't support to integrate with Enterprise Security Package (ESP). So I submitted this PR to implement it.

--- PASS: TestAccHDInsightHadoopCluster_securityProfile (1995.55s)
--- PASS: TestAccHDInsightSparkCluster_securityProfile (2476.50s)
--- PASS: TestAccHDInsightKafkaCluster_securityProfile (2082.18s)
--- PASS: TestAccHDInsightInteractiveQueryCluster_securityProfile (1806.02s)
--- PASS: TestAccHDInsightHBaseCluster_securityProfile (3006.68s)

API Reference:
https://github.com/Azure/azure-rest-api-specs/blob/a331a455fd31e022f8df07105058ae586a02ab03/specification/hdinsight/resource-manager/Microsoft.HDInsight/stable/2018-06-01-preview/cluster.json#L801

Overview:
https://docs.microsoft.com/en-us/azure/hdinsight/domain-joined/apache-domain-joined-configure-using-azure-adds

### CODE REVIEW

**Comment by grayzu** [comment]
Thanks for this PR @neil-yechenwei!

Since this is a pretty complicated scenario, can you add your tests to the examples folder so that customers can have a starting point for how to be successfully with these new properties?

**Comment by neil-yechenwei** [comment]
@grayzu , thanks for your comment. The example has been added. Thanks.

File: internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+		return
+	}
+
+	if !strings.HasPrefix(v, "ldaps://") {
+		errors = append(errors, fmt.Errorf(`%s should start with "ldaps://"`, k))
+		return
+	}
+
+	return
+}
```
Comment (manicminer) [inline internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go:21]: Can these be parsed with `url.Parse` and the scheme checked from the resulting struct?

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go [hunk 2]
```diff
@@ -2020,3 +2040,265 @@ resource "azurerm_hdinsight_hadoop_cluster" "test" {
...
+
+  depends_on = [
+    azurerm_virtual_network_dns_servers.test,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      security_profile.0.domain_user_password,
+      gateway.0.password
+    ]
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2300]: We should avoid `ignore_changes` in a test config, why are these necessary?

File: internal/services/hdinsight/hdinsight_hbase_cluster_resource_test.go [hunk 3]
```diff
@@ -1302,3 +1322,265 @@ resource "azurerm_hdinsight_hbase_cluster" "test" {
...
+
+  depends_on = [
+    azurerm_virtual_network_dns_servers.test,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      security_profile.0.domain_user_password,
+      gateway.0.password
+    ]
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_hbase_cluster_resource_test.go:1582]: As above

File: internal/services/hdinsight/hdinsight_interactive_query_cluster_resource_test.go [hunk 4]
```diff
@@ -1788,3 +1808,265 @@ resource "azurerm_hdinsight_interactive_query_cluster" "test" {
...
+
+  depends_on = [
+    azurerm_virtual_network_dns_servers.test,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      security_profile.0.domain_user_password,
+      gateway.0.password
+    ]
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_interactive_query_cluster_resource_test.go:2068]: As above

File: internal/services/hdinsight/hdinsight_kafka_cluster_resource_test.go [hunk 5]
```diff
@@ -1380,3 +1400,266 @@ resource "azurerm_hdinsight_kafka_cluster" "test" {
...
+
+  depends_on = [
+    azurerm_virtual_network_dns_servers.test,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      security_profile.0.domain_user_password,
+      gateway.0.password
+    ]
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_kafka_cluster_resource_test.go:1661]: As above

File: internal/services/hdinsight/hdinsight_spark_cluster_resource_test.go [hunk 6]
```diff
@@ -1788,3 +1808,265 @@ resource "azurerm_hdinsight_spark_cluster" "test" {
...
+
+  depends_on = [
+    azurerm_virtual_network_dns_servers.test,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      security_profile.0.domain_user_password,
+      gateway.0.password
+    ]
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_spark_cluster_resource_test.go:2068]: As above

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @neil-yechenwei, thanks for this PR. This is mostly looking good. I've made some comments inline - there's also another PR #13011 which (if merged) will make for simpler configurations.

If you can look at the below, and the tests pass, this should be good to merge. Thanks!

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 2 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2300]: Service API doesn't return these values. So we have to ignore them. Otherwise, we would encounter diff while running these TCs.

File: internal/services/hdinsight/hdinsight_hbase_cluster_resource_test.go (hunk 3 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hbase_cluster_resource_test.go:1582]: Service API doesn't return the values for these properties. So we have to ignore them. Otherwise, we would encounter diff while running these TCs.

File: internal/services/hdinsight/hdinsight_interactive_query_cluster_resource_test.go (hunk 4 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_interactive_query_cluster_resource_test.go:2068]: Service API doesn't return the values for these properties. So we have to ignore them. Otherwise, we would encounter diff while running these TCs.

File: internal/services/hdinsight/hdinsight_kafka_cluster_resource_test.go (hunk 5 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_kafka_cluster_resource_test.go:1661]: Service API doesn't return the values for these properties. So we have to ignore them. Otherwise, we would encounter diff while running these TCs.

File: internal/services/hdinsight/hdinsight_spark_cluster_resource_test.go (hunk 6 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_spark_cluster_resource_test.go:2068]: Service API doesn't return the values for these properties. So we have to ignore them. Otherwise, we would encounter diff while running these TCs.

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go [hunk 7]
```diff
@@ -2020,3 +2040,265 @@ resource "azurerm_hdinsight_hadoop_cluster" "test" {
...
+      vm_size            = "Standard_A2_V2"
+      username           = "sshuser"
+      password           = "TerrAform123!"
+      subnet_id          = azurerm_subnet.test.id
+      virtual_network_id = azurerm_virtual_network.test.id
+    }
+  }
+
+  security_profile {
+    aadds_resource_id       = "${replace(azurerm_active_directory_domain_service.test.id, "/initialReplicaSetId/${azurerm_active_directory_domain_service.test.deployment_id}", "")}"
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2283]: Updated

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 2 above)
Comment (manicminer) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2300]: There are other cases where we handle this by setting the value that's already read from state, which doesn't work for imports but is stable during normal lifecycle. e.g. the `certificate` / `certificate_password` attributes in the `hostname_configuration` block in `azurerm_api_management`

File: internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go (hunk 1 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go:21]: Updated

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go [hunk 8]
```diff
@@ -258,6 +260,17 @@ func resourceHDInsightHadoopClusterCreate(d *pluginsdk.ResourceData, meta interf
...
 	}
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+		params.Identity.UserAssignedIdentities[*params.Properties.SecurityProfile.MsiResourceID] = &hdinsight.ClusterIdentityUserAssignedIdentitiesValue{}
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go:271]: Updated

File: internal/services/hdinsight/hdinsight_hbase_cluster_resource.go [hunk 9]
```diff
@@ -204,6 +206,17 @@ func resourceHDInsightHBaseClusterCreate(d *pluginsdk.ResourceData, meta interfa
...
 	}
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+		params.Identity.UserAssignedIdentities[*params.Properties.SecurityProfile.MsiResourceID] = &hdinsight.ClusterIdentityUserAssignedIdentitiesValue{}
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hbase_cluster_resource.go:217]: Updated

File: internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go [hunk 10]
```diff
@@ -223,6 +225,17 @@ func resourceHDInsightInteractiveQueryClusterCreate(d *pluginsdk.ResourceData, m
...
 	}
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+		params.Identity.UserAssignedIdentities[*params.Properties.SecurityProfile.MsiResourceID] = &hdinsight.ClusterIdentityUserAssignedIdentitiesValue{}
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go:236]: Updated

File: internal/services/hdinsight/hdinsight_kafka_cluster_resource.go [hunk 11]
```diff
@@ -259,6 +261,16 @@ func resourceHDInsightKafkaClusterCreate(d *pluginsdk.ResourceData, meta interfa
...
 	}
 
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+		params.Identity.UserAssignedIdentities[*params.Properties.SecurityProfile.MsiResourceID] = &hdinsight.ClusterIdentityUserAssignedIdentitiesValue{}
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:271]: Updated

File: internal/services/hdinsight/hdinsight_spark_cluster_resource.go [hunk 12]
```diff
@@ -223,6 +225,17 @@ func resourceHDInsightSparkClusterCreate(d *pluginsdk.ResourceData, meta interfa
...
 	}
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+		params.Identity.UserAssignedIdentities[*params.Properties.SecurityProfile.MsiResourceID] = &hdinsight.ClusterIdentityUserAssignedIdentitiesValue{}
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_spark_cluster_resource.go:236]: Updated

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 2 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go]: OK. Thanks. Updated.

File: internal/services/hdinsight/hdinsight_hbase_cluster_resource_test.go (hunk 3 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hbase_cluster_resource_test.go]: OK. Thanks. Updated.

File: internal/services/hdinsight/hdinsight_interactive_query_cluster_resource_test.go (hunk 4 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_interactive_query_cluster_resource_test.go]: OK. Thanks. Updated.

File: internal/services/hdinsight/hdinsight_kafka_cluster_resource_test.go (hunk 5 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_kafka_cluster_resource_test.go]: OK. Thanks. Updated.

File: internal/services/hdinsight/hdinsight_spark_cluster_resource_test.go (hunk 6 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_spark_cluster_resource_test.go]: OK. Thanks. Updated.

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go [hunk 13]
```diff
@@ -2020,3 +2042,258 @@ resource "azurerm_hdinsight_hadoop_cluster" "test" {
...
+
+data "azurerm_subscription" "primary" {}
+
+resource "azurerm_role_assignment" "test" {
+  scope                = data.azurerm_subscription.primary.id
+  role_definition_name = "HDInsight Domain Services Contributor"
+  principal_id         = azurerm_user_assigned_identity.test.principal_id
+}
+
+resource "azurerm_active_directory_domain_service" "test" {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2185]: Whilst it should be fine to include an `azurerm_active_directory_domain_service` in this test, we'll probably have to think about scheduling and parallelization between this test and the test in the `domainservices` package - it's only possible to have a single domain service at any one time and they take quite a while to stand up and tear down.

File: examples/hdinsight/enterprise-security-package/main.tf [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+  depends_on = [
+    azurerm_virtual_network_dns_servers.example,
+  ]
+
+  lifecycle {
+    ignore_changes = [
+      security_profile.0.domain_user_password,
+      gateway.0.password
+    ]
+  }
```
Comment (manicminer) [inline examples/hdinsight/enterprise-security-package/main.tf:256]: We should remove this from the example too

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go [hunk 15]
```diff
@@ -258,6 +260,20 @@ func resourceHDInsightHadoopClusterCreate(d *pluginsdk.ResourceData, meta interf
...
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+
+		if params.Properties.SecurityProfile != nil && params.Properties.SecurityProfile.MsiResourceID != nil {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go:272]: `params.Properties.SecurityProfile` can still be nil here

File: internal/services/hdinsight/hdinsight_hbase_cluster_resource.go [hunk 16]
```diff
@@ -204,6 +206,20 @@ func resourceHDInsightHBaseClusterCreate(d *pluginsdk.ResourceData, meta interfa
...
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+
+		if params.Properties.SecurityProfile != nil && params.Properties.SecurityProfile.MsiResourceID != nil {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_hbase_cluster_resource.go:218]: also here

File: internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go [hunk 17]
```diff
@@ -223,6 +225,20 @@ func resourceHDInsightInteractiveQueryClusterCreate(d *pluginsdk.ResourceData, m
...
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+
+		if params.Properties.SecurityProfile != nil && params.Properties.SecurityProfile.MsiResourceID != nil {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go:237]: and here

File: internal/services/hdinsight/hdinsight_kafka_cluster_resource.go [hunk 18]
```diff
@@ -259,6 +261,19 @@ func resourceHDInsightKafkaClusterCreate(d *pluginsdk.ResourceData, meta interfa
...
 
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+
+		if params.Properties.SecurityProfile != nil && params.Properties.SecurityProfile.MsiResourceID != nil {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:272]: and here

File: internal/services/hdinsight/hdinsight_spark_cluster_resource.go [hunk 19]
```diff
@@ -223,6 +225,20 @@ func resourceHDInsightSparkClusterCreate(d *pluginsdk.ResourceData, meta interfa
...
+
+	if v, ok := d.GetOk("security_profile"); ok {
+		params.Properties.SecurityProfile = ExpandHDInsightSecurityProfile(v.([]interface{}))
+
+		params.Identity = &hdinsight.ClusterIdentity{
+			Type:                   hdinsight.ResourceIdentityTypeUserAssigned,
+			UserAssignedIdentities: make(map[string]*hdinsight.ClusterIdentityUserAssignedIdentitiesValue),
+		}
+
+		if params.Properties.SecurityProfile != nil && params.Properties.SecurityProfile.MsiResourceID != nil {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_spark_cluster_resource.go:237]: and here

File: internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+	}
+
+	ldapsUrl, err := url.Parse(v)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("parsing %q: %q", k, v))
+		return
+	}
+
+	if ldapsUrl.Scheme != "ldaps" {
+		errors = append(errors, fmt.Errorf(`%s should start with "ldaps"`, k))
```
Comment (manicminer) [inline internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go:22]:
```suggestion
		errors = append(errors, fmt.Errorf(`%s should start with "ldaps://"`, k))
```

**Review by manicminer — CHANGES_REQUESTED** [review]
@neil-yechenwei Thanks for the changes. There are still some crash opportunities, noted below, but the main thing on further thought is that `azurerm_active_directory_domain_service` has a quota of 1 per tenant or per subscription, so we'll need a way of handling that when parallelizing tests, else it'll conflict with the test in the `domainservices` package.

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2285]: Updated

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 13 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2185]: To avoid re-work, may i ask is there any suggestion from your side?

File: examples/hdinsight/enterprise-security-package/main.tf (hunk 14 above)
Comment (neil-yechenwei) [reply examples/hdinsight/enterprise-security-package/main.tf:256]: Updated

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go (hunk 15 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go:272]: So I already added nil check for `if params.Properties.SecurityProfile != nil`

File: internal/services/hdinsight/hdinsight_hbase_cluster_resource.go (hunk 16 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hbase_cluster_resource.go:218]: So I already added nil check for `if params.Properties.SecurityProfile != nil`. If I am wrong, could you provide the suggested code.

File: internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go (hunk 17 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go:237]: So I already added nil check for `if params.Properties.SecurityProfile != nil`. If I am wrong, could you provide the suggested code.

File: internal/services/hdinsight/hdinsight_kafka_cluster_resource.go (hunk 18 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:272]: So I already added nil check for `if params.Properties.SecurityProfile != nil`. If I am wrong, could you provide the suggested code.

File: internal/services/hdinsight/hdinsight_spark_cluster_resource.go (hunk 19 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_spark_cluster_resource.go:237]: So I already added nil check for `if params.Properties.SecurityProfile != nil`. If I am wrong, could you provide the suggested code.

File: internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go (hunk 20 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/validate/hdinsight_cluster_ldaps_urls.go:22]: Updated

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 13 above)
Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go]: I've updated code for your concern.

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go (hunk 15 above)
Comment (manicminer) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go:272]: Sorry my bad, misread this!

Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource.go:272]: Never mind. :)

File: internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go (hunk 13 above)
Comment (manicminer) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go]: Ah i see. Unfortunately I don't think we can consolidate the tests into one package - the tests really need to live with their respective resources.

I'm looking at the logistics of having a long-running domain service to try and compensate for the concurrency issue, but still working out how to make that work nicely with the domainservices tests themselves.

Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_hadoop_cluster_resource_test.go:2185]: OK. I separated them into their respective resources first.

**Comment by neil-yechenwei** [comment]
@manicminer , kindly ping. May I ask is there any update for above issue?

**Comment by neil-yechenwei** [comment]
@manicminer, For this test issue, is it possible that we can skip these TCs and merge this PR first. Once you figure out this kind of test issue, then enable and update the TCs.

**Comment by manicminer** [comment]
@neil-yechenwei This is challenging because having a persistent AADDS resource will break the tests for that resource. We'll leave this as-is for now, and we'll resolve separately when it becomes a problem - but can we change the AADDS SKU in this tests to `Standard`?

**Comment by neil-yechenwei** [comment]
@manicminer , I've changed the AADDS SKU in TCs to Standard.

File: internal/services/hdinsight/hdinsight_cluster_resource_test.go [hunk 21]
```diff
@@ -0,0 +1,27 @@
+package hdinsight_test
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance"
+)
+
+func TestAccHDInsightCluster(t *testing.T) {
```
Comment (manicminer) [inline internal/services/hdinsight/hdinsight_cluster_resource_test.go:9]: What's the reason for making these sequential?

Comment (neil-yechenwei) [reply internal/services/hdinsight/hdinsight_cluster_resource_test.go:9]: As you mentioned before, only one aadds is allowed per subs. So I make it sequence.

**Comment by neil-yechenwei** [comment]
@manicminer , after rerun the test cases in TeamCity, the failed test case "testAccHDInsightInteractiveQueryCluster_securityProfile" you encountered can pass now. So I think this resource is too complicated so that API isn't stable. But it isn't TF code problem.

image

**Comment by neil-yechenwei** [comment]
@manicminer , all TCs passed in TeamCity now.

image

**Review by manicminer — APPROVED** [review]
Hi @neil-yechenwei, thanks for re-running in TC. I agree this is likely to clash with the cleanup jobs due to the extended test duration. We'll likely have to revisit this to improve the stability whilst complying with the singular quota for AADDS, but in the meantime this should be good to merge 👍
